### ISSUE TITLE
Flutter Plugin Template generates 2 instances of the Plugin class in Android

### ISSUE THREAD

**Issue by Vanethos** [flutter/flutter#48802]
# Problem:
The `onAttachedToEngine` method in the template project creates a new instance of the same class instead of referencing it.
This means that there will be 2 instances of the `XPlugin` class running at the same time, one handling the `MethodChannel`and another that isn't.

# Description:

When creating a new plugin via the template `flutter create -t plugin test_plugin`, the following file is generated:

```kotlin
public class TestingPlugin: FlutterPlugin, MethodCallHandler {
  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "testing_plugin")
    channel.setMethodCallHandler(TestingPlugin());
  }
  
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "testing_plugin")
      channel.setMethodCallHandler(TestingPlugin())
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
  }
}
```

As seen in `onAttachedToEngine`, a new instance of the plugin is created via `channel.setMethodCallHandler(TestingPlugin());` though we are already in the context of the Plugin.

In this case, isn't it safer to reference the same class using `this` so that users don't have reference issues? Also this way we will only have 1 instance of the class up and running.

**Example**:
* In `onAttachedToEngine` we set a specific value.
* However, a new instance of a `TestPlugin` is created and the Method Channel is assigned to it
* In this case, the second instance of `TestPlugin` will not receive the `onAttachedToEngine` so the value is not assigned
* When we receive a command via the Method Channel we will be referencing a null object in the second instance of `TestPlugin`

```kotlin
public class MinimizeAppPlugin(): FlutterPlugin, MethodCallHandler, ActivityAware {
  public var activity : Activity? = null

  override fun onDetachedFromActivity() {}

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    onAttachedToActivity(binding)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    this.activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {}


  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding?.getFlutterEngine()?.getDartExecutor(), "minimize_app")
    channel.setMethodCallHandler(MinimizeAppPlugin())
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {}

  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "minimize_app")
      val plugin = MinimizeAppPlugin();
      channel.setMethodCallHandler(plugin)
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    this.activity?.moveTaskToBack(true)
    result.success(null)
  }
}
```

In the above code, inside `onMethodCall`, `activity` will be null.

However changing the `onAttachedToEngine` to use `this`, solves the problem:

```kotlin
override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding?.getFlutterEngine()?.getDartExecutor(), "minimize_app")
    channel.setMethodCallHandler(this)
  }
```

**Comment by jmagman** [comment]
\cc @blasten  https://github.com/flutter/flutter/pull/41666/files#diff-ac3da909bb5c97ea4e17ed460f74b7ebR21

**Comment by blasten** [comment]
wdyt @matthew-carroll ?

**Comment by matthew-carroll** [comment]
Hi @Vanethos - WRT to instantiating plugin classes during registration in V2, rather than invoking static registration like V1, this change was very much intentional. The plugin class, itself, now merely represents the registration with a specific `FlutterEngine`. Plugin authors can still implement their plugin in a singleton manner by utilizing a static implementation, but had we implemented plugins the other way around then the plugin authors that didn't want singleton behavior would have needed to hack a solution.

However, I do believe you have discovered an issue with the plugin template. The following implementation does appear to be incorrect:

```kotlin
override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
  val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "testing_plugin")
  channel.setMethodCallHandler(TestingPlugin());
}
```

As you suggest, it does look like the appropriate implementation should be:

```kotlin
override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
  val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "testing_plugin")
  channel.setMethodCallHandler(this);
}
```

We should make this change.

It would probably also be a good idea to explicitly null out the handler in detachment in the template:

```kotlin
override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
  channel.setMethodCallHandler(null);
}
```

**Comment by Vanethos** [comment]
Hi @matthew-carroll ! Thanks for the explanation!

Have to say that I also really like this solution and it was very easy to implement in a new small plugin that I made.

About your changes I have just one question:

In order to do `channel.setMethodCallHandler(null)`, we have to get a hold of the `MethodChannel`. Is it enough to create another `MethodChannel` object using the same initialization: 

```dart
var channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "testing_plugin")
```

Or do we have to access the previous `channel` variable created on `onAttachedToEngine`?

**Comment by matthew-carroll** [comment]
@Vanethos to execute `channel.setMethodCallHandler(null)` you'll need to use the same `channel` instance, so in that case you should hold onto it in a local variable in your plugin class. You can assign that variable in `onAttachedToEngine()` and then you can invoke `setMethodCallHandler(null)` and unassign the variable in `onDetachedFromEngine()`.

**Comment by Vanethos** [comment]
I'll add it ASAP.

**Comment by v7lin** [comment]
#49105 I have same issue

**Comment by Vanethos** [comment]
@matthew-carroll in addition to adding the `MethodChannel` as a private variable, I added the following comment: 

```
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
```

**Comment by jessndaa** [comment]
attach method call with this
```kotlin
  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding.flutterEngine.dartExecutor, "ussd")
    channel.setMethodCallHandler(this)
  }
```

**Comment by Vanethos** [comment]
Closed via #48806
